src/supervisor.tex

```diff
@@ -368,12 +368,20 @@ \subsection{Supervisor Interrupt Registers ({\tt sip} and {\tt sie})}
 \label{siereg}
 \end{figure}
 
-An interrupt \textit{i} will be taken if bit \textit{i} is set in both
-{\tt sip} and {\tt sie}, and if supervisor-level interrupts are globally
-enabled.
-Supervisor-level interrupts are globally enabled if the hart's current
-privilege mode is less than S, or if the current privilege mode is S
-and the SIE bit in the {\tt sstatus} register is set.
+An interrupt~\textit{i} will trap to S-mode if both of the
+following are true:
+(a)~either the current privilege mode is S and the SIE bit in the
+{\tt sstatus} register is set, or the current privilege mode has less
+privilege than S-mode; and
+(b)~bit~\textit{i} is set in both {\tt sip} and {\tt sie}.
+
+These conditions for an interrupt trap to occur must be evaluated in a bounded
+amount of time from when an interrupt becomes pending in {\tt sip}, and must
+also be evaluated immediately following the execution of an SRET instruction
+or an explicit write to {\tt sip} or {\tt sie}.
```

Comment:
Some questions and observations that all reflect a certain unease over the loosy-goosy-ness of interrupt recognition.

1)  What about before or after an explicit read of sip, i.e. should one be able to directly observe a pending enabled interrupt that remains to be taken?  Further, currently it seems like one could repeatedly read sip and see an enabled interrupt continuing to be pending for a "bounded amount of time".   Although this appears to be the direct intended implication since how else can one know that an interrupt is pending in sip.

2a)  An enabled interrupt can be long pending and then disabled by a sip write without any guarantee of ever being taken.  But then, when the interrupt is re-enabled, there is now an immediate guarantee of being taken.

2b)  It seems a little weird that an enabled interrupt can remain pending for an unknown (but bounded) amount of time, but then by simply writing sie (even if the enable status remains unchanged), the interrupt is now guaranteed to immediately be taken.  This smacks of a bit of logical arbitrariness in having a change to unrelated pending or enable bits force a behavior on taking of a given interrupt.

3a)  Setting an enabled sip bit via an explicit write must immediately be evaluated and taken right after the write instruction - and hence before it could ever be observed via a sip read.  But when an enabled sip bit is set externally, then it isn't required to be taken before or even after it can be observed via a sip read.

3b)  It seems somewhat inconsistent that if an enabled interrupt becomes pending an instant before an explicit sip write, then it must be taken immediately after the write.  Whereas if an interrupt becomes pending an instant after an explicit write, then it can remain pending for a "bounded" amount of time.

Obviously John's proposed definition was clean and simple, but much tighter.  What about just adding the idea that an enabled interrupt cannot explicitly be observed as pending, i.e. via a sip read?  Sort of like quantum mechanics - there are no status guarantees until you make an observation.  There may or may not be a long pending interrupt.  But once you try to observe a pending interrupt, you must only be able to observe a disabled pending interrupt (and an enabled pending interrupt must be taken by or before the point of observation).

This could amount to simply adding to the end of the proposed sentence "..., and before execution of an explicit read of sip."

Maybe also add that an enabled pending interrupt must be taken before a sip/sie write (especially since you're already requiring a tight sync between the instruction's execution and post-recognition of pending enabled interrupts)?  Which leads to "... or immediately before and after execution of an explicit read or write of sip or sie."

In any case, your call.  Sorry for the long rambling unease.

Reply:
I _want_ to allow the possibility that software reads `sip` and detects the presence of an interrupt that might've been taken but hasn't yet been--exactly the opposite of what you're getting at.  As was discussed on #590, many of us don't perceive this behavior to be a problem.  Allowing this oddity means that reads of `sip` don't need to be special-cased to avoid an inconsistency that doesn't really matter.

You mentioned on #590 an x86 behavior that you understandably want to mimic, and this proposal does satisfy that constraint.

Another prominent architecture seems to have taken the same tack as I'm proposing.  In their parlance, writes of sip/sie are context-synchronizing w.r.t. interrupts.  In @jhauser-us's proposal, _all_ instructions are context-synchronizing w.r.t. interrupts.